pkg/products/grafana/reconciler.go

```diff
@@ -0,0 +1,320 @@
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, fmt.Sprintf("Failed to reconcile %s ns", operatorNamespace), err)
+		return phase, err
+	}
+
+	err = r.createSecrets(ctx, client, installation, &grafanav1alpha1.Grafana{})
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, fmt.Sprintf("Failed to reconcile %s ns", productNamespace), err)
+		return phase, err
+	}
+
+	phase, err = r.reconcileSubscription(ctx, client, installation, productNamespace, operatorNamespace)
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, fmt.Sprintf("Failed to reconcile %s subscription", constants.ThreeScaleSubscriptionName), err)
+		return phase, err
+	}
+
+	phase, err = r.reconcileComponents(ctx, client, installation, &grafanav1alpha1.Grafana{})
+	if err != nil || phase != integreatlyv1alpha1.PhaseCompleted {
+		events.HandleError(r.recorder, installation, phase, "Failed to create components", err)
+		return phase, err
+	}
+
+	return integreatlyv1alpha1.PhaseCompleted, nil
+}
+func (r *Reconciler) createSecrets(ctx context.Context, client k8sclient.Client, installation *integreatlyv1alpha1.RHMI, cr *grafanav1alpha1.Grafana) error {
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "grafana-k8s-proxy",
+			Namespace: r.Config.GetOperatorNamespace(),
+		},
+	}
+	_, err := controllerutil.CreateOrUpdate(ctx, client, secret, func() error {
+		owner.AddIntegreatlyOwnerAnnotations(secret, installation)
+		if secret.Data == nil {
+			secret.Data = map[string][]byte{}
+		}
+		secret.Data["session_secret"] = []byte(PopulateSessionProxySecret())
+		return nil
+	})
+
+	if err != nil {
+		logrus.Errorf("failed to create grafana proxy secret, %s: %w", secret, err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) reconcileComponents(ctx context.Context, client k8sclient.Client, installation *integreatlyv1alpha1.RHMI, cr *grafanav1alpha1.Grafana) (integreatlyv1alpha1.StatusPhase, error) {
+	r.logger.Info("reconciling grafana custom resource")
+
+	var annotations = map[string]string{}
+	annotations["service.alpha.openshift.io/serving-cert-secret-name"] = "grafana-k8s-tls"
+
+	var serviceAccountAnnotations = map[string]string{}
+	serviceAccountAnnotations["serviceaccounts.openshift.io/oauth-redirectreference.primary"] = "{\"kind\":\"OAuthRedirectReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"Route\",\"name\":\"grafana-route\"}}"
+
+	grafana := &grafanav1alpha1.Grafana{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "grafana",
+			Namespace: r.Config.GetOperatorNamespace(),
+		}, Spec: grafanav1alpha1.GrafanaSpec{
+			Config: grafanav1alpha1.GrafanaConfig{
+				Log: &grafanav1alpha1.GrafanaConfigLog{
+					Mode:  "console",
+					Level: "warn",
+				},
+				Auth: &grafanav1alpha1.GrafanaConfigAuth{
+					DisableLoginForm:   &[]bool{false}[0],
```

Comment:
this is nuts, is this because it has to be a pointer to a bool?

Reply:
I think so, the error I get when using just true/false is 
cannot convert false (untyped bool constant) to *bool
Is there a better way to do it ?

Reply:
no, this is fine, just looks odd! :)